lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java

```diff
@@ -116,38 +115,16 @@ public void writeBlock(ColumnChunk columnChunk)
         // write values
         primitiveValueWriter.write(columnChunk.getBlock());
 
-        if (columnChunk.getDefLevelIterables().isEmpty()) {
-            // write definition levels for flat data types
-            Block block = columnChunk.getBlock();
-            if (!block.mayHaveNull()) {
-                for (int position = 0; position < block.getPositionCount(); position++) {
-                    definitionLevelWriter.writeInteger(maxDefinitionLevel);
-                }
-            }
-            else {
-                for (int position = 0; position < block.getPositionCount(); position++) {
-                    byte isNull = (byte) (block.isNull(position) ? 1 : 0);
-                    definitionLevelWriter.writeInteger(maxDefinitionLevel - isNull);
-                    currentPageNullCounts += isNull;
-                }
-            }
-            valueCount += block.getPositionCount();
-        }
-        else {
-            // write definition levels for nested data types
-            Iterator<Integer> defIterator = DefLevelWriterProviders.getIterator(ImmutableList.<DefLevelWriterProvider>builder()
-                    .addAll(columnChunk.getDefLevelIterables())
-                    .add(DefLevelWriterProviders.of(columnChunk.getBlock(), maxDefinitionLevel))
-                    .build());
-            while (defIterator.hasNext()) {
-                int next = defIterator.next();
-                definitionLevelWriter.writeInteger(next);
-                if (next != maxDefinitionLevel) {
-                    currentPageNullCounts++;
-                }
-                valueCount++;
-            }
-        }
+        List<DefLevelWriterProvider> defIterables = ImmutableList.<DefLevelWriterProvider>builder()
+                .addAll(columnChunk.getDefLevelIterables())
+                .add(DefLevelWriterProviders.of(columnChunk.getBlock(), maxDefinitionLevel))
+                .build();
+        DefLevelWriterProvider.ValuesCount valuesCount = defIterables.get(0).getDefinitionLevelWriter(
```

Comment:
Instead of making `DefLevelWriterProvider` responsible for creating a sucessor, create hierarchy explicitly here from leaf to head